airflow/models/backfill.py

```diff
@@ -59,4 +89,97 @@ class BackfillDagRun(Base):
+    dag_run = relationship("DagRun")
+
+    __table_args__ = (
+        UniqueConstraint("backfill_id", "dag_run_id", name="ix_bdr_backfill_id_dag_run_id"),
+        ForeignKeyConstraint(
+            [backfill_id],
+            ["backfill.id"],
+            name="bdr_backfill_fkey",
+            ondelete="cascade",
+        ),
+        ForeignKeyConstraint(
+            [dag_run_id],
+            ["dag_run.id"],
+            name="bdr_dag_run_fkey",
+            ondelete="set null",
+        ),
+    )
+
+
+def _create_backfill(
+    *,
+    dag_id: str,
+    from_date: DateTime,
+    to_date: DateTime,
+    max_active_runs: int,
+    reverse: bool,
+    dag_run_conf: dict | None,
+) -> Backfill | None:
+    with create_session() as session:
+        serdag = session.get(SerializedDagModel, dag_id)
+        if not serdag:
+            raise NotFound(f"Could not find dag {dag_id}")
+
+        num_active = session.scalar(
+            select(func.count()).where(Backfill.dag_id == dag_id, Backfill.completed_at.is_(None))
+        )
+        if num_active > 0:
+            raise AlreadyRunningBackfill(
+                f"Another backfill is running for dag {dag_id}. "
+                f"There can be only one running backfill per dag."
+            )
+
+        br = Backfill(
+            dag_id=dag_id,
+            from_date=from_date,
+            to_date=to_date,
+            max_active_runs=max_active_runs,
+            dag_run_conf=dag_run_conf,
+        )
+        session.add(br)
+        session.commit()
+
+        dag = serdag.dag
+        depends_on_past = any(x.depends_on_past for x in dag.tasks)
+        if depends_on_past:
+            if reverse is True:
+                raise ValueError(
+                    "Backfill cannot be run in reverse when the dag has tasks where depends_on_past=True"
+                )
+
+        backfill_sort_ordinal = 0
+        dagrun_info_list = dag.iter_dagrun_infos_between(from_date, to_date)
+        if reverse:
+            dagrun_info_list = reversed([x for x in dag.iter_dagrun_infos_between(from_date, to_date)])
+        for info in dagrun_info_list:
+            backfill_sort_ordinal += 1
+            log.info("creating backfill dag run %s dag_id=%s backfill_id=%s, info=", dag.dag_id, br.id, info)
+            dr = None
+            try:
+                dr = dag.create_dagrun(
+                    triggered_by=DagRunTriggeredByType.BACKFILL,
+                    execution_date=info.logical_date,
+                    data_interval=info.data_interval,
+                    start_date=timezone.utcnow(),
+                    state=DagRunState.QUEUED,
+                    external_trigger=False,
+                    conf=br.dag_run_conf,
+                    run_type=DagRunType.BACKFILL_JOB,
+                    creating_job_id=None,
+                    session=session,
+                )
+            except Exception:
+                dag.log.exception("something failed")
```

Comment:
We can probably do a better log message here.

Reply:
lol yeah

Reply:
i will do in followup.

Reply:
improving the message in https://github.com/apache/airflow/pull/42530 since it had to be rebased anyway